File: app/src/main/java/com/github/pockethub/ui/repo/RepositoryListFragment.java
/*
 * Copyright (c) 2015 PocketHub
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.pockethub.ui.repo;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.support.v4.content.Loader;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.ListView;

import com.afollestad.materialdialogs.MaterialDialog;
import com.alorma.github.sdk.bean.dto.response.Repo;
import com.alorma.github.sdk.bean.dto.response.User;
import com.github.kevinsawicki.wishlist.SingleTypeAdapter;
import com.github.kevinsawicki.wishlist.ViewFinder;
import com.github.pockethub.R;
import com.github.pockethub.ThrowableLoader;
import com.github.pockethub.persistence.AccountDataManager;
import com.github.pockethub.ui.HeaderFooterListAdapter;
import com.github.pockethub.ui.ItemListFragment;
import com.github.pockethub.ui.user.OrganizationSelectionListener;
import com.github.pockethub.ui.user.OrganizationSelectionProvider;
import com.github.pockethub.ui.user.UserViewActivity;
import com.github.pockethub.util.AvatarLoader;
import com.github.pockethub.util.InfoUtils;
import com.google.inject.Inject;

import java.util.Collections;
import java.util.List;
import java.util.concurrent.atomic.AtomicReference;

import static com.github.pockethub.Intents.EXTRA_USER;
import static com.github.pockethub.RequestCodes.REPOSITORY_VIEW;
import static com.github.pockethub.ResultCodes.RESOURCE_CHANGED;
import static java.util.Locale.US;

/**
 * Fragment to display a list of {@link Repo} instances
 */
public class RepositoryListFragment extends ItemListFragment<Repo>
    implements OrganizationSelectionListener {

    @Inject
    private AccountDataManager cache;

    @Inject
    private AvatarLoader avatars;

    private final AtomicReference<User> org = new AtomicReference<>();

    private RecentRepositories recentRepos;

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);

        User org = this.org.get();
        if (org != null)
            outState.putParcelable(EXTRA_USER, org);
    }

    @Override
    protected void configureList(Activity activity, ListView listView) {
        super.configureList(activity, listView);

        listView.setDividerHeight(0);
        updateHeaders(items);
    }

    @Override
    public void onDetach() {
        if (getActivity() != null && getActivity() instanceof OrganizationSelectionProvider) {
            OrganizationSelectionProvider selectionProvider = (OrganizationSelectionProvider) getActivity();
            selectionProvider.removeListener(this);
        }

        super.onDetach();
    }

    @Override
    public void onOrganizationSelected(final User organization) {
        User previousOrg = org.get();
        int previousOrgId = previousOrg != null ? previousOrg.id : -1;
        org.set(organization);

        if (recentRepos != null)
            recentRepos.saveAsync();

        // Only hard refresh if view already created and org is changing
        if (previousOrgId != organization.id) {
            Activity activity = getActivity();
            if (activity != null)
                recentRepos = new RecentRepositories(activity, organization);

            refreshWithProgress();
        }
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        Activity activity = getActivity();
        User currentOrg = null;

        if (getActivity() instanceof OrganizationSelectionProvider)
            currentOrg = ((OrganizationSelectionProvider) activity)
                .addListener(this);

        if (getArguments() != null && getArguments().containsKey("org"))
            currentOrg = getArguments().getParcelable("org");

        if (currentOrg == null && savedInstanceState != null)
            currentOrg = savedInstanceState.getParcelable(EXTRA_USER);
        org.set(currentOrg);
        if (currentOrg != null)
            recentRepos = new RecentRepositories(activity, currentOrg);

        setEmptyText(R.string.no_repositories);

        super.onActivityCreated(savedInstanceState);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        // Refresh if the viewed repository was (un)starred
        if (requestCode == REPOSITORY_VIEW && resultCode == RESOURCE_CHANGED) {
            forceRefresh();
            return;
        }

        super.onActivityResult(requestCode, resultCode, data);
    }

    @Override
    public void onListItemClick(ListView list, View v, int position, long id) {
        Repo repo = (Repo) list.getItemAtPosition(position);
        if (recentRepos != null)
            recentRepos.add(repo);

        startActivityForResult(RepositoryViewActivity.createIntent(repo),
            REPOSITORY_VIEW);
    }

    @Override
    public boolean onListItemLongClick(ListView list, View v, int position,
        long itemId) {
        if (!isUsable())
            return false;

        final Repo repo = (Repo) list.getItemAtPosition(position);
        if (repo == null)
            return false;

        MaterialDialog.Builder builder = new MaterialDialog.Builder(getActivity())
                .title(InfoUtils.createRepoId(repo));
        final MaterialDialog[] dialogHolder = new MaterialDialog[1];

        View view = getActivity().getLayoutInflater().inflate(
            R.layout.repo_dialog, null);
        ViewFinder finder = new ViewFinder(view);

        final User owner = repo.owner;
        avatars.bind(finder.imageView(R.id.iv_owner_avatar), owner);
        finder.setText(R.id.tv_owner_name, getString(R.string.navigate_to_user, owner.login));
        finder.onClick(R.id.ll_owner_area, new OnClickListener() {
            public void onClick(View v) {
                dialogHolder[0].dismiss();
                viewUser(owner);
            }
        });

        if ((recentRepos != null) && (recentRepos.contains(repo))) {
            finder.find(R.id.divider).setVisibility(View.VISIBLE);
            finder.find(R.id.ll_recent_repo_area).setVisibility(View.VISIBLE);
            finder.onClick(R.id.ll_recent_repo_area, new OnClickListener() {
                public void onClick(View v) {
                    dialogHolder[0].dismiss();
                    recentRepos.remove(repo);
                    refresh();
                }
            });
        }

        builder.customView(view, false);
        MaterialDialog dialog = builder.build();
        dialogHolder[0] = dialog;
        dialog.setCanceledOnTouchOutside(true);
        dialog.show();

        return true;
    }

    private void viewUser(User user) {
        if (org.get().id != user.id)
            startActivity(UserViewActivity.createIntent(user));
    }

    @Override
    public void onStop() {
        super.onStop();

        if (recentRepos != null)
            recentRepos.saveAsync();
    }

    private void updateHeaders(final List<Repo> repos) {
        HeaderFooterListAdapter<?> rootAdapter = getListAdapter();
        if (rootAdapter == null)
            return;

        DefaultRepositoryListAdapter adapter = (DefaultRepositoryListAdapter) rootAdapter
            .getWrappedAdapter();
        adapter.clearHeaders();

        if (repos.isEmpty())
            return;

        // Add recent header if at least one recent repository
        Repo first = repos.get(0);
        if (recentRepos.contains(first))
            adapter.registerHeader(first, getString(R.string.recently_viewed));

        // Advance past all recent repositories
        int index;
        Repo current = null;
        for (index = 0; index < repos.size(); index++) {
            Repo repository = repos.get(index);
            if (recentRepos.contains(repository.id))
                current = repository;
            else
                break;
        }

        if (index >= repos.size())
            return;

        if (current != null)
            adapter.registerNoSeparator(current);

        // Register header for first character
        current = repos.get(index);
        char start = Character.toLowerCase(current.name.charAt(0));
        adapter.registerHeader(current,
            Character.toString(start).toUpperCase(US));

        char previousHeader = start;
        for (index = index + 1; index < repos.size(); index++) {
            current = repos.get(index);
            char repoStart = Character.toLowerCase(current.name.charAt(0));
            if (repoStart <= start)
                continue;

            // Don't include separator for the last element of the previous
            // character
            if (previousHeader != repoStart)
                adapter.registerNoSeparator(repos.get(index - 1));

            adapter.registerHeader(current, Character.toString(repoStart)
                .toUpperCase(US));
            previousHeader = repoStart;
            start = repoStart++;
        }

        // Don't include separator for last element
        adapter.registerNoSeparator(repos.get(repos.size() - 1));
    }

    @Override
    public Loader<List<Repo>> onCreateLoader(int id, final Bundle args) {
        return new ThrowableLoader<List<Repo>>(getActivity(), items) {

            @Override
            public List<Repo> loadData() throws Exception {
                User org = RepositoryListFragment.this.org.get();
                if (org == null)
                    return Collections.emptyList();

                List<Repo> repos = cache.getRepos(org,
                    isForceRefresh(args));
                Collections.sort(repos, recentRepos);
                updateHeaders(repos);
                return repos;
            }
        };
    }

    @Override
    protected SingleTypeAdapter<Repo> createAdapter(List<Repo> items) {
        return new DefaultRepositoryListAdapter(getActivity()
            .getLayoutInflater(),
            items.toArray(new Repo[items.size()]), org);
    }

    @Override
    protected int getErrorMessage(Exception exception) {
        return R.string.error_repos_load;
    }
}

File: app/src/main/java/com/github/pockethub/android/ui/repo/RepositoryListFragment.kt
/*
 * Copyright (c) 2015 PocketHub
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.pockethub.android.ui.repo

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.afollestad.materialdialogs.MaterialDialog
import com.github.pockethub.android.Intents.EXTRA_USER
import com.github.pockethub.android.ItemListHandler
import com.github.pockethub.android.ListFetcher
import com.github.pockethub.android.R
import com.github.pockethub.android.RequestCodes.REPOSITORY_VIEW
import com.github.pockethub.android.ResultCodes.RESOURCE_CHANGED
import com.github.pockethub.android.persistence.AccountDataManager
import com.github.pockethub.android.ui.base.BaseFragment
import com.github.pockethub.android.ui.item.repository.RepositoryHeaderItem
import com.github.pockethub.android.ui.item.repository.RepositoryItem
import com.github.pockethub.android.ui.user.OrganizationSelectionListener
import com.github.pockethub.android.ui.user.OrganizationSelectionProvider
import com.github.pockethub.android.ui.user.UserViewActivity
import com.github.pockethub.android.util.AvatarLoader
import com.github.pockethub.android.util.InfoUtils
import com.github.pockethub.android.util.ToastUtils
import com.meisolsson.githubsdk.model.Repository
import com.meisolsson.githubsdk.model.User
import com.xwray.groupie.Item
import com.xwray.groupie.OnItemClickListener
import com.xwray.groupie.OnItemLongClickListener
import io.reactivex.Observable
import io.reactivex.Single
import kotlinx.android.synthetic.main.fragment_item_list.view.*
import kotlinx.android.synthetic.main.repo_dialog.view.*
import java.util.Locale.US
import java.util.concurrent.atomic.AtomicReference
import javax.inject.Inject

/**
 * Fragment to display a list of [Repository] instances
 */
class RepositoryListFragment : BaseFragment(), OrganizationSelectionListener {

    @Inject
    protected lateinit var cache: AccountDataManager

    @Inject
    protected lateinit var avatars: AvatarLoader

    private lateinit var itemListHandler: ItemListHandler

    private lateinit var listFetcher: ListFetcher<Repository>

    private val org = AtomicReference<User>()

    private var recentRepos: RecentRepositories? = null

    private val errorMessage: Int
        get() = R.string.error_repos_load

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.fragment_item_list, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        itemListHandler = ItemListHandler(
            view.list,
            view.empty,
            lifecycle,
            activity,
            OnItemClickListener(this::onItemClick),
            OnItemLongClickListener(this::onItemLongClick)
        )

        listFetcher = ListFetcher(
            view.swipe_item,
            lifecycle,
            itemListHandler,
            { t -> ToastUtils.show(activity, errorMessage)},
            this::loadData,
            this::createItem
        )

        listFetcher.onDataLoaded = this::onDataLoaded
        itemListHandler.setEmptyText(R.string.no_repositories)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)

        val org = this.org.get()
        if (org != null) {
            outState.putParcelable(EXTRA_USER, org)
        }
    }

    override fun onDetach() {
        if (activity != null && activity is OrganizationSelectionProvider) {
            val selectionProvider = activity as OrganizationSelectionProvider?
            selectionProvider!!.removeListener(this)
        }

        super.onDetach()
    }

    override fun onOrganizationSelected(organization: User) {
        val previousOrg = org.get()
        val previousOrgId = if (previousOrg != null) previousOrg.id() else -1
        org.set(organization)

        if (recentRepos != null) {
            recentRepos!!.saveAsync()
        }

        // Only hard refresh if view already created and org is changing
        if (previousOrgId != organization.id()) {
            val activity = activity
            if (activity != null) {
                recentRepos = RecentRepositories(activity, organization)
            }

            listFetcher.forceRefresh()
        }
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        val activity = activity
        var currentOrg: User? = null

        if (getActivity() is OrganizationSelectionProvider) {
            currentOrg = (activity as OrganizationSelectionProvider).addListener(this)
        }

        if (arguments != null && arguments!!.containsKey("org")) {
            currentOrg = arguments!!.getParcelable("org")
        }

        if (currentOrg == null && savedInstanceState != null) {
            currentOrg = savedInstanceState.getParcelable(EXTRA_USER)
        }

        org.set(currentOrg)
        if (currentOrg != null) {
            recentRepos = RecentRepositories(activity, currentOrg)
        }

        super.onActivityCreated(savedInstanceState)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        // Refresh if the viewed repository was (un)starred
        if (requestCode == REPOSITORY_VIEW && resultCode == RESOURCE_CHANGED) {
            listFetcher.forceRefresh()
            return
        }

        super.onActivityResult(requestCode, resultCode, data)
    }

    private fun onItemClick(item: Item<*>, view: View) {
        if (item is RepositoryItem) {
            val repo = item.repo
            if (recentRepos != null) {
                recentRepos!!.add(repo)
            }

            startActivityForResult(RepositoryViewActivity.createIntent(repo),
                REPOSITORY_VIEW)
        }
    }

    private fun onItemLongClick(item: Item<*>, view: View): Boolean {
        if (!isAdded) {
            return false
        }

        if (item is RepositoryItem) {
            val repo = item.repo

            val builder = MaterialDialog.Builder(activity!!).title(InfoUtils.createRepoId(repo))
            val dialogHolder = arrayOfNulls<MaterialDialog>(1)

            val dialogView = layoutInflater.inflate(R.layout.repo_dialog, null)

            val owner = repo.owner()
            avatars.bind(dialogView.iv_owner_avatar, owner)
            dialogView.tv_owner_name.text = getString(R.string.navigate_to_user, owner!!.login())
            dialogView.ll_owner_area.setOnClickListener { v1 ->
                dialogHolder[0]!!.dismiss()
                viewUser(owner)
            }

            if (recentRepos != null && recentRepos!!.contains(repo)) {
                dialogView.divider.visibility = View.VISIBLE
                val recentRepoArea = dialogView.ll_recent_repo_area
                recentRepoArea.visibility = View.VISIBLE
                recentRepoArea.setOnClickListener { v1 ->
                    dialogHolder[0]!!.dismiss()
                    recentRepos!!.remove(repo)
                    listFetcher.refresh()
                }
            }

            builder.customView(dialogView, false)
            val dialog = builder.build()
            dialogHolder[0] = dialog
            dialog.setCanceledOnTouchOutside(true)
            dialog.show()

            return true
        }

        return false
    }

    private fun viewUser(user: User) {
        if (org.get().id() !== user.id()) {
            startActivity(UserViewActivity.createIntent(user))
        }
    }

    override fun onStop() {
        super.onStop()

        if (recentRepos != null) {
            recentRepos!!.saveAsync()
        }
    }

    private fun updateHeaders(repos: MutableList<Item<*>>) {
        if (repos.isEmpty()) {
            return
        }

        // Add recent header if at least one recent repository
        val first = (repos[0] as RepositoryItem).repo
        if (recentRepos!!.contains(first)) {
            repos.add(0, RepositoryHeaderItem(getString(R.string.recently_viewed)))
        }

        // Advance past all recent repositories
        var index = 0
        while (index < repos.size) {
            val item = repos[index]
            if (item is RepositoryItem) {
                val repository = item.repo
                if (!recentRepos!!.contains(repository.id()!!)) {
                    break
                }
            }
            index++
        }

        if (index >= repos.size) {
            return
        }

        // Register header for first character
        var current = (repos[index] as RepositoryItem).repo
        var start = Character.toLowerCase(current.name()!![0])
        repos.add(index, RepositoryHeaderItem(Character.toString(start).toUpperCase(US)))

        index += 1
        while (index < repos.size) {
            current = (repos[index] as RepositoryItem).repo
            val repoStart = Character.toLowerCase(current.name()!![0])
            if (repoStart <= start) {
                index++
                continue
            }

            repos.add(index, RepositoryHeaderItem(Character.toString(repoStart).toUpperCase(US)))
            start = repoStart
            index++
        }
    }

    private fun loadData(forceRefresh: Boolean): Single<List<Repository>> {
        val org = this.org.get() ?: return Single.just(emptyList())

        return Single.fromCallable { cache.getRepos(org, forceRefresh) }
            .flatMap { repos ->
                Observable.fromIterable(repos)
                    .sorted(recentRepos!!)
                    .toList()
            }
    }

    private fun createItem(item: Repository): Item<*> {
        return RepositoryItem(item, org.get())
    }

    protected fun onDataLoaded(newItems: MutableList<Item<*>>): MutableList<Item<*>> {
        updateHeaders(newItems)
        return newItems
    }
}
